x/validator-preference/keeper/msg_server.go

```diff
@@ -0,0 +1,66 @@
+package keeper
+
+import (
+	"context"
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/osmosis-labs/osmosis/v12/x/validator-preference/types"
+)
+
+type msgServer struct {
+	keeper *Keeper
+}
+
+// NewMsgServerImpl returns an implementation of the MsgServer interface
+// for the provided Keeper.
+func NewMsgServerImpl(keeper *Keeper) types.MsgServer {
+	return &msgServer{
+		keeper: keeper,
+	}
+}
+
+var _ types.MsgServer = msgServer{}
+
+func (server msgServer) SetValidatorSetPreference(goCtx context.Context, msg *types.MsgSetValidatorSetPreference) (*types.MsgSetValidatorSetPreferenceResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	// check if a user already have a validator-set created
+	_, found := server.keeper.GetValidatorSetPreference(ctx, msg.Delegator)
+	if found {
+		// TODO;
```

Comment:
@ValarDragon want to run through this logic with you. 

I'm not sure if its good to have undelege + delegate logic in one flow